### PR TITLE
Add dashboard

### LINKED ISSUE #75: Create dashboard
## User Story

**As a** user -  
**I want to** have a dashboard where I can see what's currently going on on the platform -  
**In order to** quickly engage in spontaneous communication.

## Conditions of satisfaction
- [x] The dashboard should contain placeholders for the following features:
  - [x] setting the own status (e.g. dropdown, top left) [#76]
  - [x] list of all users (not only contacts) filtered by status (left, 20% width) [#31]
  - [x] global chat, filtered by status (left-center, 40%) [#77]
  - [x] last activities  (center-right, 20%)
    - [x] creation (top) [#79]
    - [x] feed [#80]
  - [x] engagement proposals (right, 20%) / meeting invitations list [#37]
- [x] the dashboard should be displayed in a sensible manner on mobile screens
  - [x] a design for that is developed 
  - [x] the design is approved by PO
  - [x] the design is implemented

## Implementation Note

Filling the placeholders mentioned above ***is not*** part of this issue.

The design might look something like this (note that the content of this draft ***is not*** what is required to be implemented here. I rather put it for optical guidance).

image

### PR DESCRIPTION
**Pull request by sirkrypt0**

### CODE REVIEW

File: app/views/home/dashboard.html+mobile.erb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+<div id="dashboard">
+  <div id="status">
+    <%= render 'status' %>
```
Comment (konradh) [inline app/views/home/dashboard.html+mobile.erb]:
```suggestion
    <%= render 'status_dropdown' %>
```
The new status dropdown is in another partial than the one before.

File: app/views/home/dashboard.html+mobile.erb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  </div>
+
+  <div id="content">
+    <div class="tab-content">
+      <div id="users" class="tab-pane fade show active">
+        <%= render 'user_list' %>
```
Comment (konradh) [inline app/views/home/dashboard.html+mobile.erb:9]:
```suggestion
        <%= render 'status_dropdown' %>
        <%= render 'user_list' %>
```
Only render the status selection in the user list.

File: app/views/home/dashboard.html+mobile.erb [hunk 3]
```diff
@@ -0,0 +1,43 @@
+<div id="dashboard">
+  <div id="status">
+    <%= render 'status' %>
+  </div>
```
Comment (konradh) [inline app/views/home/dashboard.html+mobile.erb:4]:
```suggestion
```

Comment (konradh) [inline app/views/home/dashboard.html+mobile.erb]: This is the wrong place if the status should only displayed in the user list.

**Review by konradh — CHANGES_REQUESTED** [review]

File: app/views/home/_status_dropdown.html.erb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<%= form_for(current_user, url: update_status_user_path(current_user), method: :patch) do |f| %>
+  <div class="status-dropdown">
+    <%= f.label :current_status %>
+    <%= f.select(:current_status, current_user.select_status_list, {}, {onchange: "this.form.submit()", class: "form-control"}) %>
+  </div>
+<% end %>
```
Comment (felixauringer) [inline app/views/home/_status_dropdown.html.erb]: newline missing

File: app/views/home/_users.html.erb [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    </div>
+    <div class="col user-name">
+      Person Personson
+    </div>
+    <div class="user-meeting d-flex">
+      <span><i class="fa fa-phone fa-flip-horizontal"></i></span>
+    </div>
+  </div>
+
+</div>
```
Comment (felixauringer) [inline app/views/home/_users.html.erb]: newline missing

File: app/views/home/dashboard.html+mobile.erb (hunk 3 above)
Comment (konradh) [inline app/views/home/dashboard.html+mobile.erb]:
```suggestion
```

File: app/views/home/dashboard.html+mobile.erb (hunk 2 above)
Comment (konradh) [inline app/views/home/dashboard.html+mobile.erb]:
```suggestion
        <%= render 'status' %>
        <%= render 'user_list' %>
```

File: app/assets/stylesheets/desktop.scss [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+// $light: #dd6108;
+// $dark: #dd6108;
+
+// https://github.com/twbs/bootstrap-rubygem
+@import 'bootstrap';
+// https://github.com/bokmann/font-awesome-rails
+@import 'font-awesome';
+
+// app/assets/stylesheets/notes.scss
+@import 'notes';
```
Comment (frcroth) [inline app/assets/stylesheets/desktop.scss:44]: ?

File: app/assets/stylesheets/mobile.scss [hunk 7]
```diff
@@ -43,4 +43,5 @@ $success: #00911f;
 // app/assets/stylesheets/notes.scss
 @import 'notes';
```
Comment (frcroth) [inline app/assets/stylesheets/mobile.scss:44]: Can we remove this to see what happens?

File: app/models/user.rb [hunk 8]
```diff
@@ -8,12 +8,12 @@ class User < ApplicationRecord
   # The dependent: option allows to specify that associated records should be deleted when the owner is deleted
   # https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html >> Deleting from Associations
   has_many :notes, dependent: :delete_all
+  has_many :activities, dependent: :delete_all
   has_and_belongs_to_many :contacts,
                           class_name: 'User',
-                          foreign_key: 'user_id',
```
Comment (frcroth) [inline app/models/user.rb]: Why was this deleted?

File: app/views/home/_status.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+  <div class="dropdown">
+    <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdownMenuButton" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+      Set status
+    </button>
+    <div class="dropdown-menu" aria-labelledby="dropdownMenuButton">
+      <a class="dropdown-item" href="#">available</a>
+      <a class="dropdown-item" href="#">working</a> <!-- should be done dynamically-->
+    </div>
+  </div>
+</div>
```
Comment (frcroth) [inline app/views/home/_status.html.erb]: Can this file be deleted?

File: app/views/home/_users.html.erb [hunk 10]
```diff
@@ -0,0 +1,119 @@
+<h1 class="center-heading"><%= I18n.t('dashboard.user_list') %></h1>
```
Comment (frcroth) [inline app/views/home/_users.html.erb:1]: Can this file be deleted?

File: app/views/home/dashboard.html.erb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+<!--<meta http-equiv="refresh" content="2" />-->
+
```
Comment (frcroth) [inline app/views/home/dashboard.html.erb]:
```suggestion
```

File: app/views/home/index.html.erb [hunk 12]
```diff
@@ -1,37 +1,13 @@
...
 <div class="jumbotron">
-  <h1 class="display-4">Connections Portal</h1>
+      <h1 class="display-4">Connections Portal</h1>
   <p class="lead">A digital tool for organizing and arranging networks, relationships and human connections.</p>
   <hr>
   <p>We hope you enjoy using it!
-    <a href="https://github.com/hpi-swt2/connections-portal/issues/new" target="_blank">Give feedback!</a></p>
+        <a href="https://github.com/hpi-swt2/connections-portal/issues/new" target="_blank">Give feedback!</a></p>
   <p class="lead" id="registration">
     <%# fa_icon helper: https://github.com/bokmann/font-awesome-rails %>
```
Comment (frcroth) [inline app/views/home/index.html.erb:8]: Comment? Remove

File: app/views/home/index.html.erb [hunk 13]
```diff
@@ -1,37 +1,13 @@
+<!--<meta http-equiv="refresh" content="2" />-->
 <div class="jumbotron">
-  <h1 class="display-4">Connections Portal</h1>
+      <h1 class="display-4">Connections Portal</h1>
   <p class="lead">A digital tool for organizing and arranging networks, relationships and human connections.</p>
   <hr>
   <p>We hope you enjoy using it!
```
Comment (frcroth) [inline app/views/home/index.html.erb]: Should be localised

File: app/views/users/show.html.erb [hunk 14]
```diff
@@ -24,7 +24,7 @@
 </table>
 
 <% if user_signed_in? && current_user.id == @user.id %>
-  <%= render 'status_form' %>
+  <%= render 'home/status_dropdown' %>
```
Comment (frcroth) [inline app/views/users/show.html.erb]: Maybe that file should be placed in another position when it is not exclusive to home. Like "common" for example.

File: spec/features/dashboard_spec.rb [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  it 'can change status by selecting item in status dropdown' do
+    user.update(current_status: 'working')
+    target_status = 'free_for_chat'
+    page.execute_script('document.getElementById("user_current_status").querySelector("option[value=\"' +
+                          target_status + '\"]").selected = true')
+    # required since we change the node programmatically
+    page.execute_script('document.getElementById("user_current_status").onchange()')
+    sleep(0.1) # required, as we need to wait for the post request to be sent and processed
```
Comment (frcroth) [inline spec/features/dashboard_spec.rb]: Is this actually required?

File: spec/features/user_profile_spec.rb [hunk 16]
```diff
@@ -9,11 +9,20 @@
...
+  it 'can change status by selecting item in status dropdown' do
+    user.update(current_status: 'working')
+    target_status = 'free_for_chat'
+    page.execute_script('document.getElementById("user_current_status").querySelector("option[value=\"' +
+                          target_status + '\"]").selected = true')
+    # required since we change the node programmatically
+    page.execute_script('document.getElementById("user_current_status").onchange()')
+    sleep(0.1) # required, as we need to wait for the post request to be sent and processed
+    user.reload
+    expect(user.current_status).to eq(target_status)
```
Comment (frcroth) [inline spec/features/user_profile_spec.rb]: Code duplication. Maybe use helper?

File: app/assets/stylesheets/mobile.scss (hunk 7 above)
Comment (sirkrypt0) [reply app/assets/stylesheets/mobile.scss:44]: removing it should be fine

File: spec/requests/users/user_status_update_spec.rb [hunk 17]
```diff
@@ -6,10 +6,10 @@
   let(:status_params) { { user: { current_status: 'working' } } }
```
Comment (felixauringer) [inline spec/requests/users/user_status_update_spec.rb:6]: should `I18n` be used here?

File: spec/features/user_profile_spec.rb [hunk 18]
```diff
@@ -25,6 +34,10 @@
     expect(page).not_to have_select('user[current_status]')
   end
 
+  it 'contains the "free for chat" status' do
+    expect(page).to have_select('user[current_status]', with_options: ['free for chat'])
```
Comment (felixauringer) [inline spec/features/user_profile_spec.rb:30]: maybe we should add all options here? (with `I18n`)

File: spec/features/user_profile_spec.rb [hunk 19]
```diff
@@ -9,11 +9,20 @@
...
+  end
+
+  it 'can change status by selecting item in status dropdown' do
+    user.update(current_status: 'working')
+    target_status = 'free_for_chat'
+    page.execute_script('document.getElementById("user_current_status").querySelector("option[value=\"' +
+                          target_status + '\"]").selected = true')
+    # required since we change the node programmatically
+    page.execute_script('document.getElementById("user_current_status").onchange()')
+    sleep(0.1) # required, as we need to wait for the post request to be sent and processed
```
Comment (felixauringer) [inline spec/features/user_profile_spec.rb]: You could have a look at how we have done this for the activities. A sleep should not be required.

File: spec/features/dashboard_spec.rb (hunk 15 above)
Comment (felixauringer) [reply spec/features/dashboard_spec.rb]: have a look at the activities

File: spec/features/user_profile_spec.rb (hunk 16 above)
Comment (felixauringer) [reply spec/features/user_profile_spec.rb]: +1

File: app/views/users/index.html.erb [hunk 20]
```diff
@@ -4,15 +4,13 @@
...
-    <th width="100"></th>
   </tr>
   </thead>
-
   <tbody>
-  <% @users.each do |user| %>
+  <% User.all.each do |user| %>
     <tr>
       <td><%= user.email %></td>
-      <td align="center"><%= link_to 'Show', user %></td>
```
Comment (felixauringer) [inline app/views/users/index.html.erb:15]: Is there still a way to navigate to the user's profile page?

File: app/views/layouts/_navbar.html.erb [hunk 21]
```diff
@@ -23,14 +26,14 @@
...
+            <%= link_to I18n.t('navigation.account_settings'), edit_user_registration_path(current_user), class: 'dropdown-item' %>
+          </div>
+        </li>
+        <li class="nav-item">
+            <%= link_to (fa_icon 'sign-out', text: 'Log Out'), destroy_user_session_path, method: :delete, class: 'nav-link' %>
+        </li>
+      <% end %>
+    </ul>
+  </div>
 </nav>
```
Comment (felixauringer) [inline app/views/layouts/_navbar.html.erb]: missing newline

File: app/views/layouts/_navbar.html.erb [hunk 22]
```diff
@@ -10,6 +10,9 @@
       <li class="nav-item">
         <%= link_to (fa_icon 'sticky-note', text: 'Notes'), notes_path, class: 'nav-link' %>
```
Comment (felixauringer) [inline app/views/layouts/_navbar.html.erb]: `I18n`

Comment (felixauringer) [reply app/views/layouts/_navbar.html.erb]: (all over this file)

File: app/views/home/index.html.erb (hunk 12 above)
Comment (felixauringer) [reply app/views/home/index.html.erb:8]: +1

File: app/views/home/index.html.erb [hunk 23]
```diff
@@ -1,37 +1,13 @@
+<!--<meta http-equiv="refresh" content="2" />-->
```
Comment (felixauringer) [inline app/views/home/index.html.erb]: Can be removed?

File: app/views/home/_users.html.erb (hunk 10 above)
Comment (felixauringer) [reply app/views/home/_users.html.erb:1]: I think it is still used because the filtering is not implemented yet

File: app/views/home/_user_list.html.erb [hunk 24]
```diff
@@ -0,0 +1,18 @@
+<div>
```
Comment (felixauringer) [inline app/views/home/_user_list.html.erb]: What's the difference to `app/views/users/index.html.erb`?

File: app/views/users/index.html.erb [hunk 25]
```diff
@@ -4,15 +4,13 @@
   <thead>
   <tr>
     <th>Email</th>
-    <th width="100"></th>
   </tr>
   </thead>
-
   <tbody>
-  <% @users.each do |user| %>
+  <% User.all.each do |user| %>
```
Comment (felixauringer) [inline app/views/users/index.html.erb]: Ahhhhh, database querying in the view!

File: spec/helpers/activity_helper_spec.rb [hunk 26]
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
Comment (frcroth) [inline spec/helpers/activity_helper_spec.rb]: Delete file

File: app/assets/stylesheets/mobile.scss [hunk 27]
```diff
@@ -43,4 +43,5 @@ $success: #00911f;
...
 // app/assets/stylesheets/home.scss
-@import 'home';
\ No newline at end of file
+@import 'home';
```
Comment (felixauringer) [inline app/assets/stylesheets/mobile.scss]: not sure if they are necessary, especially the desktop variant?

**Review by felixauringer — CHANGES_REQUESTED** [review]
I think I saw some lines that rubocop would not accept ;)

File: spec/features/dashboard_spec.rb (hunk 15 above)
Comment (sirkrypt0) [reply spec/features/dashboard_spec.rb]: > Is this actually required?

Haven't you added this? ^^

File: app/views/home/index.html.erb (hunk 12 above)
Comment (sirkrypt0) [reply app/views/home/index.html.erb:8]: We haven't added this one, maybe it's meant as a reference for the use of the icons.

File: app/views/users/index.html.erb (hunk 25 above)
Comment (sirkrypt0) [reply app/views/users/index.html.erb]: Just copied what was on the home page before :P

File: app/assets/stylesheets/mobile.scss (hunk 27 above)
Comment (sirkrypt0) [reply app/assets/stylesheets/mobile.scss]: Yes, should be removed as otherwise we have that Stylesheet loaded in all mobile views.

File: spec/features/dashboard_spec.rb (hunk 15 above)
Comment (frcroth) [reply spec/features/dashboard_spec.rb]: Dunno, anyway at that point the js option was not added to the spec I think.

Comment (frcroth) [reply spec/features/dashboard_spec.rb]: Still required it seems

File: spec/requests/users/user_status_update_spec.rb (hunk 17 above)
Comment (frcroth) [reply spec/requests/users/user_status_update_spec.rb:6]: I don't think the request params are localized

File: app/assets/stylesheets/desktop.scss (hunk 6 above)
Comment (frcroth) [reply app/assets/stylesheets/desktop.scss:44]: Ignore for now

File: spec/features/user_profile_spec.rb (hunk 19 above)
Comment (frcroth) [reply spec/features/user_profile_spec.rb]: Ignore for now
